tango/executors/multicore_executor.py

```diff
@@ -0,0 +1,237 @@
+import logging
+import os
+import subprocess
+import time
+from tempfile import NamedTemporaryFile
+from typing import Dict, List, Optional, OrderedDict, Set, TypeVar
+
+from tango.executor import Executor
+from tango.step import Step
+from tango.step_graph import StepGraph
+from tango.workspace import StepState, Workspace
+
+logger = logging.getLogger(__name__)
+
+T = TypeVar("T")
+
+
+class MulticoreExecutor(Executor):
+    """
+    A ``MulticoreExecutor`` runs the steps in parallel and caches their results.
+    """
+
+    def __init__(
+        self,
+        workspace: Workspace,
+        include_package: Optional[List[str]] = None,
+        parallelism: int = 1,
+        num_tries_to_sync_states: int = 3,
+        wait_seconds_to_sync_states: int = 3,
+    ) -> None:
+        self.workspace = workspace
+        self.include_package = include_package
+        self.parallelism = parallelism
+
+        # Perhaps there's a better way to do this without these being passed as args.
+        self._num_tries_to_sync_states = num_tries_to_sync_states
+        self._wait_seconds_to_sync_states = wait_seconds_to_sync_states
+
+    def execute_step_graph(self, step_graph: StepGraph, run_name: Optional[str] = None) -> Set[str]:
+        """
+        Execute a :class:`tango.step_graph.StepGraph`.
+        """
+
+        _running: OrderedDict[str, subprocess.Popen] = OrderedDict({})
+        _failed: Set[str] = set()
+        _queued_steps: List[str] = []
+
+        def _sync_step_states() -> Dict[str, StepState]:
+            """
+            Update the StepState info.
+            Although, this is not really elegant. The issue is as follows: The main multicore executor process
+            queues a step, and starts step execution in a different process. If we try to read the StepState
+            before that process has had time to update the StepState, the Workspace will throw the out of sync
+            error (IOError: process should be running but it's considered incomplete...).
```

Comment:
Yes, I can solve it this way, but there's an edge case that I ran into: 

[This](https://github.com/allenai/tango/blob/main/test_fixtures/package/steps.py#L40) step is supposed to be non-deterministic, but according to Tango, it's not. It's for our test case, so we can change it, but users can write what they want. 

And we have [this](https://github.com/allenai/tango/blob/main/test_fixtures/experiment/random.jsonnet) config with: 
```
"steps": {
        "rand_string1": {"type": "random_string"},
        "rand_string2": {"type": "random_string"},
        ...
```

When calling `workspace.step_info(step)` to sync the states for each step, both these steps will be considered the same (as according to Tango, they are deterministic), so if `rand_string1` has been kicked off, and we don't check it, we will run into an error when checking `rand_string2`'s step state. With the current implementation, it will gracefully find that both steps have finished, but if I only check `_running`, it will fail. 

We may not want to support this edge case since, technically, there should be no need to run a deterministic step twice. 

The other option (which makes more sense, I guess) is to explicitly check if any steps are the same (but with different names), in which case, we warn the user, and remove the duplicate step from the run's config, and update the dependencies. Or just plainly fail without running anything if we find a duplicate step.

Reply:
In some ways, this mimics the issue that will occur with concurrent runs, so I think it still makes sense to keep the current implementation.